### DISCUSSION TITLE
Enabling a parent parameter from a child

### DISCUSSION THREAD

**Discussion by ceraolo** [OpenModelica/OpenModelica#12761] [General]
I would like to have inheritance such that Child1 is able to enable (through a boolean parameter) parent's parameter `parParent`, while `parParent` is always enabled in Child2. Therefore in Child2 the enabling boolean parameter should not be displayed.

I've prepared the following code:
```modelica
package P
  partial model Parent
    parameter Real parParent=1  annotation(
      Dialog(enable = enableParParent));
  end Parent;

  model Child1
    extends Parent;
    parameter Boolean enableParParent = true;
  end Child1;

  model Child2
    extends Parent;
  end Child2;

  model Test
  Child1 child1(enableParParent=true)
                annotation(
      Placement(transformation(origin = {-30, 50}, extent = {{-10, -10}, {10, 10}})));
  Child2 child2 annotation(
      Placement(transformation(origin = {-30, 10}, extent = {{-10, -10}, {10, 10}})));
  end Test;
end P;
```
It works in Dymola but has the following weaknesses:
- it is considered incorrect by Dymola in pedantic mode
- OpenModelica issues an error when analysing Parent and the enabling mechanism from Child1 does not work.

Does anyone know a correct Modelica way to reach my purpose?
Or must I resign myself to move ParParam to children (which, among others, would imply constraints on the position in which the parameter appears in user's parameter's window)?

Thanks anyone.

**Comment by casella** [comment]
The `Parent` model is invalid because it uses `enableParParent` without ever defining it. Of course this is not allowed in Modelica, I wonder how can Dymola ever accept such a model, even without pedantic mode.

I guess what you need to write is:
```modelica
package P
  partial model Parent
    parameter Real parParent=1  annotation(
      Dialog(enable = enableParParent));
   parameter Boolean enableParParent;
  end Parent;

  model Child1
    extends Parent;
  end Child1;

  model Child2
    extends Parent(final enableParParent = true);
  end Child2;

  model Test
  Child1 child1(enableParParent=true)
                annotation(
      Placement(transformation(origin = {-30, 50}, extent = {{-10, -10}, {10, 10}})));
  Child2 child2 annotation(
      Placement(transformation(origin = {-30, 10}, extent = {{-10, -10}, {10, 10}})));
  end Test;
end P;
```

**Reply by ceraolo** [reply]
> I wonder how can Dymola ever accept such a model, even without pedantic mode.

In reality, I expected this. If the analysis is done when flat modelica has already been created, enableParParent is defined, and this would be in priciple acceptable for me since Parent is declared as partial.

Thank you, @casella for your fixed code (and very fast answer)

**Reply by casella** [reply]
In some sense your child model were ok, but that's not how Modelica works, every model (also the partial ones) must be valid.
